Source/Entities/Validation/ValidationRules.cst

```diff
@@ -38,6 +39,13 @@ namespace <%=NameSpace%>.Validation
       [NonSerialized()]
       private Dictionary<string, List<ValidationRuleInfo>> _rulesList;
 
+      /// <summary>
+	  /// Object associated with the common/generated rules list.
+	  /// </summary>>
+      private static ConcurrentDictionary<string, ConcurrentDictionary<string, List<ValidationRuleInfo>>> _commonRulesList;

```

Comment:
Not entirely sure if these need to be ConcurrentDictionaries...  But seemed safer to do so.

Reply:
ConcurrentDictionary was not introduced until .net 4 it looks like:
http://dotnetpattern.com/csharp-concurrentdictionary

Would it be possible to do a version check on the version being generated here, and swap it out for a more basic dictionary style if we are generating for a version older than 4?

Reply:
Should I abandon the whole thing if the target framework can't support ConcurrentDictionary?  I am worried about the thread safety if it's just a static Dictionary.  they are notorious for that.

Reply:
I think it is a good concept.  I have never run into the memory issues you are having, but conceptually it seems like a really good improvement.

If you do not believe you can effectively and safely create a system that will work with older versions, then I would suggest you try creating versioned copies of the rule system file/files.  Leave the old system alone for older versions of .net, and generate the newer files for newer versions.

@niemyjski has made multiple suggestions that we stop supporting older versions of .net for new features.  I have conflicting feelings on the idea.  But I think the main reason no movement has been made on this suggestion is that no one has had the time to think through the idea and administratively archive off an older copy of the code base that is dedicated to the earlier versions and is no longer maintained.

Reply:
I have removed the concurrent dictionary and replaced it with manual locking.

I also have used some crude local unit tests to attempt to test the thread safety and it looks ok.

Reply:
I have also ensured it builds when .Net 3 is selected in the parameters.

Reply:
I really really think we should drop everything but .NET 4.8 and only support the latest versions of everything (e.g., EntLib). It's not worth anyones time supporting old options that is not easy to test, nor supported by Microsoft anymore.

Reply:
@niemyjski  do you want to make a copy of the nettiers code base, and mark it as legacy supporting back to .net 2 ?

We'll have to drop the support in stages though for several reasons.  It looks like Microsoft supports back to .net 4.5.2.  nettiers though only offers .net 4.5 as the most current generation option, and only offers visual studio 2013 as the most current environment to target.  Plus, enterprise library is no longer supported at all, but nettiers requires it, although we do support the latest version of it now.

Reply:
It might only take an hour or two to update the visual studio and .net versions.
But removing the enterprise library requirement is going to take a lot more work.  I'm only about half way there so far.  We still have error handling and data access that use it heavily.

Reply:
So what needs to happen with regards to this PR?

As an FYI - I am running this in production.

Reply:
I have just realised I failed to push the most recent changes.  I will do so now.